src/config.rs

```diff
@@ -0,0 +1,200 @@
+    config_path: Option<PathBuf>,
+
+    /// Default registry to use if none is specified
+    default_registry: Option<String>,
+
+    /// List of registries
+    registries: HashMap<String, RegistryUri>,
+}
+
+impl Config {
+    /// Create a new configuration with default values
+    /// # Arguments
+    /// * `cwd` - Starting directory to search for the configuration file
+    ///
+    pub fn new(cwd: Option<&Path>) -> miette::Result<Self> {
+        match Self::locate_config(cwd) {
+            Some(config_path) => Self::new_from_config_file(&config_path),
+            None => Ok(Self {
+                config_path: None,
+                default_registry: None,
+                registries: HashMap::new(),
+            }),
+        }
+    }
+
+    /// Resolve the registry URI from the configuration
+    ///
+    /// # Arguments
+    /// * `registry` - The registry name or URI to resolve
+    ///
+    /// # Returns
+    /// The resolved registry URI
+    pub fn registry_or_default(&self, registry: &Option<String>) -> miette::Result<RegistryUri> {
+        match registry {
+            Some(registry) => {
+                match RegistryUri::from_str(registry) {
+                    Ok(uri) => Ok(uri),
+                    Err(_) => self.lookup_registry(registry),
+                }
+            }
+            None => match &self.default_registry {
+                Some(default_registry) => self
+                    .registries
+                    .get(default_registry)
+                    .cloned()
+                    .ok_or_else(|| {
+                        miette!("no registry provided (using --registry) and no default registry in .buffrs/config.toml")
+                    }),
+                None => bail!("no registry provided and no default registry found"),
+            },
+        }
+    }
+
+    /// Lookup a registry by name
+    ///
+    /// # Arguments
+    /// * `name` - Name of the registry to lookup
+    ///
+    /// # Returns
+    /// The registry URI
+    pub fn lookup_registry(&self, name: &str) -> miette::Result<RegistryUri> {
+        self.registries.get(name).cloned().ok_or_else(|| {
+            miette!(
+                "registry '{}' not found in {}",
+                name,
+                self.config_path
+                    .clone()
+                    .unwrap_or("config file".into())
+                    .display()
+            )
+        })
+    }
+
+    /// Locate the configuration file in the current directory or any parent directories
+    ///
+    /// # Arguments
+    /// * `cwd` - Starting directory to search for the configuration file
+    ///
+    /// # Returns
+    /// Some(PathBuf) if the configuration file is found, None otherwise
+    fn locate_config(cwd: Option<&Path>) -> Option<PathBuf> {
+        if let Some(cwd) = cwd {
+            let mut current_dir = cwd.to_owned();
+
+            loop {
+                let config_path = current_dir.join(".config/buffrs/config.toml");
```

Comment:
Slight misunderstanding, the config should be placed in a user home directory, i.e. `~/.config/buffrs`. The exact path can be customized, so there's a helper crate to resolve the actual path:
```suggestion
                use etcetera::BaseStrategy;

                let config_path = etcetera::choose_base_strategy()?.config_dir().join("buffrs/config.toml");
```

and in `Cargo.toml`
```toml
etcetera = "0.8.0"
```

Then, you can actually remove the looping logic and the whole `locate_config` function and just use the above snippet in `Config::new()`.

Reply:
@tomkarw : placing the config in the user's home directory actually defies the very purpose of this PR. In our org, we need to pre-configure the registry URIs on a per-repo basis. So this definitely needs to live with the repo, not the user.

Without this feature, I prefer to close the PR.

Let me know your opinion -- if we agree then I'll address your other concerns below.

Reply:
Hi @dgehriger i am sorry that I was unresponsive with regards to your PRs and I'm happy to jump on a call if you want to reopen them to figure out how we can merge your work.

For this specific change I do agree that the configuration file would need to be included in the repository itself rather than the home directory.

Also, naively asked, how does this affect dependency resolution?

Reply:
@mara-schulke : no problem, we all have our projects and timelines. I'd actually love to get on a call with you (maybe next week ?) to sync up. We are using your wonderful tool inside our organization, but it's true that I had to fork it to make sure it matches our workflows. It would be wonderful if, instead of us maintaining a diverging version of `buffrs`, we could bring this together. But it's not a hard requirement on my side.

Can you drop me a mail on dgehriger@globusmedical.com and suggest a time that suits you for a video call during (ideally sometimes between 9 - 12 am Central European Time -- I'm located close to you, in Lausanne, Switzerland)?

Reply:
@tomkarw : since your suggestion was regarding _~/.config/..._ and doesn't apply to repository configuration files, I've reverted the change in the latest commit.